src/engine/TPA.cc

```diff
@@ -1691,6 +1694,40 @@ InvalidityWitness TransitionSystemNetworkManager::computeInvalidityWitness() con
     return InvalidityWitness::fromErrorPath(path, graph);
 }
 
+ValidityWitness TransitionSystemNetworkManager::computeValidityWitness() {
+    if (not isTransitionSystemDAG(graph)) { return {}; }
+    auto vertices = graph.getVertices();
+    assert(vertices.size() >= 3);
+    TermUtils utils(logic);
+    TimeMachine timeMachine(logic);
+    TermUtils::substitutions_map subs;
+    std::vector<ValidityWitness> witnesses;
+    ValidityWitness::definitions_t definitions;
+
+    for (auto node : vertices) {
+        auto vertex = node;
+        if (vertex == graph.getEntry() || vertex == graph.getExit()) continue;
+
+        auto graphVars = utils.predicateArgsInOrder(graph.getStateVersion(vertex));
+        vec<PTRef> unversionedVars;
+        auto systemVars = networkMap.at(vertex).solver->getStateVars(0);
+
+        for (std::size_t i = 0; i < graphVars.size(); ++i) {
+            unversionedVars.push(timeMachine.getUnversioned(graphVars[i]));
+            subs.insert({systemVars[i], unversionedVars.last()});
+        }
+        getNode(vertex).solver->resetInitialStates(getNode(vertex).trulySafe);
+        auto [res, explanation] = queryTransitionSystem(networkMap.at(vertex));
```

Comment:
Maybe add a note here that this uses the query that was set in the main solving algorithm?

Reply:
Done